internal/internal_task_handlers.go

```diff
@@ -1495,11 +1495,10 @@ func (i *cadenceInvoker) heartbeatAndScheduleNextRun(details []byte) error {
 			i.Unlock()
 
 			// Log the error outside the lock.
-			if err != nil {
-				// If the error is a canceled error do not log, as this is expected.
-				if _, ok := err.(*CanceledError); !ok {
-					i.logger.Error("Failed to send heartbeat", zap.Error(err), zap.String(tagWorkflowType, i.workflowType), zap.String(tagActivityType, i.activityType))
-				}
+			// If the error is a canceled error do not log, as this is expected.
+			var canceledErr *CanceledError
+			if canceledErr != nil && !errors.As(err, &canceledErr) {
```

Comment:
canceledErr is always nil and err is checked for nil inside errors.As.

Reply:
also, maybe put this into a method and add a test so you are 100% sure about the behaviour. 
Something like 
i.processHearbeatError(err)
and then add unit test with zaptest.Logger.